Sources/QueryResult.swift

```diff
@@ -36,54 +52,104 @@ public final class QueryResult {
-            }
-
-            rows.append(ResultRow(columnValues: values))
+        for rowIndex in 0..<self.numberOfRows {
+            rows.append(self.readResultRowAtIndex(rowIndex))
         }
 
         return rows
     }()
+
+    private func readResultRowAtIndex(rowIndex: Int32) -> ResultRow {
+        var values = [Any?]()
+        values.reserveCapacity(Int(self.numberOfColumns))
+
+        for columnIndex in 0..<self.numberOfColumns {
+            values.append(readColumnValueAtIndex(columnIndex, rowIndex: rowIndex))
+        }
+
+        return ResultRow(columnValues: values, queryResult: self)
+    }
+
+    private func readColumnValueAtIndex(columnIndex: Int32, rowIndex: Int32) -> Any? {
+        guard PQgetisnull(self.resultPointer, rowIndex, columnIndex) == 0 else { return nil }
+
+        let firstBytePointer = PQgetvalue(self.resultPointer, rowIndex, columnIndex)
+
+        guard let type = self.typesForColumnIndexes[Int(columnIndex)] else {
+            let length = Int(PQgetlength(self.resultPointer, rowIndex, columnIndex))
+            return readBytesStartingAtPointer(firstBytePointer, length: length)
+        }
+
+        switch type {
+        case .Boolean: return UnsafeMutablePointer<Bool>(firstBytePointer).memory
+        case .Int16: return swapInt16Bytes(UnsafeMutablePointer<Int16>(firstBytePointer).memory)
+        case .Int32: return swapInt32Bytes(UnsafeMutablePointer<Int32>(firstBytePointer).memory)
+        case .Int64: return swapInt64Bytes(UnsafeMutablePointer<Int64>(firstBytePointer).memory)
+//        case .SingleFloat: return parseDouble()
+//            case .DoubleFloat: return parseDouble(bytes)
+        case .Text: return String.fromCString(firstBytePointer)!
+            default: return nil
+        }
+    }
+
+    private func readBytesStartingAtPointer(pointer: UnsafeMutablePointer<Int8>, length: Int) -> [Int8] {
+        var pointer = pointer
+        var bytes: [Int8] = []
+        bytes.reserveCapacity(length)
+
+        for _ in 0..<length {
+            bytes.append(pointer.memory)
+            pointer = pointer.advancedBy(1)
+        }
+
+        return bytes
+    }
 }
 
 public struct ResultRow {
-    public let columnValues: [ColumnValue]
-}
+    public let columnValues: [Any?]
+    unowned let queryResult: QueryResult
 
-public enum ColumnValue {
-    case Boolean(Bool?)
-    case Data([UInt8]?)
-    case DoubleType(Double?)
-    case Integer(Int?)
-    case Text(String?)
+    subscript(columnName: String) -> Any? {
+        get {
+            guard let index = queryResult.columnIndexesForNames[columnName] else { return nil }
+            return columnValues[index]
+        }
+    }
 }
 
 enum ColumnType: UInt32 {
     case Boolean = 16
-    case SingleFloat = 700
-    case DoubleFloat = 701
     case Int64 = 20
     case Int16 = 21
     case Int32 = 23
     case Text = 25
+    case SingleFloat = 700
+    case DoubleFloat = 701
 }
 
+private func swapInt16Bytes(input: Int16) -> Int16 {
```

Comment:
Since we are defining these, and not using an underlying implementation's signature, can we give them more descriptive names? `16BitNetworkToHostTransformation` or something similar?